include/mem.h

```diff
@@ -159,11 +109,104 @@ constexpr static INLINE uint32_t host_to_le(uint32_t val) {
 	return val;
 }
 
+constexpr static INLINE uint64_t host_to_le(uint64_t val)
+{
+	return val;
+}
+
 #endif
 
-static INLINE void var_write(Bit8u * var, Bit8u val) {
-	host_writeb(var, val);
+constexpr static INLINE uint8_t le_to_host(uint8_t val)
+{
+	return host_to_le(val);
+}
+
+constexpr static INLINE int16_t le_to_host(int16_t val)
+{
+	return host_to_le(val);
+}
+
+constexpr static INLINE uint16_t le_to_host(uint16_t val)
+{
+	return host_to_le(val);
+}
+
+constexpr static INLINE uint32_t le_to_host(uint32_t val)
+{
+	return host_to_le(val);
+}
+
+constexpr static INLINE uint64_t le_to_host(uint64_t val)
+{
+	return host_to_le(val);
+}
+
+// Read, write, and add using 16-bit words
+static INLINE uint16_t host_readw(const uint8_t *arr)
+{
+	uint16_t val;
+	memcpy(reinterpret_cast<void *>(&val),
+	       reinterpret_cast<const void *>(arr), sizeof(val));
+	// array sequence was DOS little-endian, so convert value to host-type
+	return le_to_host(val);
 }
+static INLINE void host_writew(uint8_t *arr, uint16_t val)
+{
+	// Convert the host-type value to little-endian before filling array
+	val = host_to_le(val);
+	memcpy(reinterpret_cast<void *>(arr),
+	       reinterpret_cast<const void *>(&val), sizeof(val));
+}
+static INLINE void host_addw(uint8_t *arr, const uint16_t &incr)
+{
+	const uint16_t val = host_readw(arr) + incr;
+	host_writew(arr, val);
+}
+
+// Read, write, and add using 32-bit double-words
+static INLINE uint32_t host_readd(const uint8_t *arr)
+{
+	uint32_t val;
+	memcpy(reinterpret_cast<void *>(&val),
+	       reinterpret_cast<const void *>(arr), sizeof(val));
+	// array sequence was DOS little-endian, so convert value to host-type
+	return le_to_host(val);
+}
+static INLINE void host_writed(uint8_t *arr, uint32_t val)
+{
+	// Convert the host-type value to little-endian before filling array
+	val = host_to_le(val);
+	memcpy(reinterpret_cast<void *>(arr),
+	       reinterpret_cast<const void *>(&val), sizeof(val));
+}
+static INLINE void host_addd(uint8_t *arr, const uint32_t &incr)
```

Comment:
I like how this function is type-safe and user is allowed to increment only by `uin32_t`. This tiny detail might get lost on readers though ("why second paramter is a reference when it doesn't need to be?"), so maybe a comment with explanation?

"Second paramter is const-reference to disallow accidental type promotion" (?)

Reply:
Yes, that was the intent - but on second though, the destination type (16bit or 32bit) is explicitly controlled by the function name: `host_add`_**d**_(...) vs `host_add`**_w_**(...); so there should be no accidents there.

As for the incrementing value, I guess it doesn't need to be reference, given it's already capped both numerically and has the same type-size as the destination; so no chance of over-or-under reading.  I've switched it passing by value now, which is probably a bit faster too.